src/dataflow/src/render/reduce.rs

```diff
@@ -114,34 +133,33 @@ where
                     |_expr| None,
                     move |row| {
                         let temp_storage = RowArena::new();
-                        // Ensure the packer is clear, and does not reflect
-                        // columns from prior rows that may have errored.
-                        row_packer.clear();
 
-                        // First, evaluate the key selector expressions.
+                        // Unpack only the demanded columns.
                         let mut datums_local = datums.borrow();
-                        datums_local.extend(row.iter().take(columns_needed));
-                        for expr in group_key_clone.iter() {
-                            match expr.eval(&datums_local, &temp_storage) {
-                                Ok(val) => row_packer.push(val),
-                                Err(e) => {
-                                    return Some(Err(e.into()));
-                                }
-                            }
+                        let mut row_iter = row.iter();
+                        for skip in demand.iter() {
+                            datums_local.push((&mut row_iter).nth(*skip).unwrap());
                         }
 
-                        // Second, evaluate the value selector expressions.
-                        let key = row_packer.finish_and_reuse();
-                        for aggr in aggregates_clone.iter() {
-                            match aggr.expr.eval(&datums_local, &temp_storage) {
-                                Ok(val) => {
-                                    row_packer.push(val);
-                                }
-                                Err(e) => {
-                                    return Some(Err(e.into()));
-                                }
-                            }
-                        }
+                        // Evaluate the key expressions.
+                        row_packer.clear();
+                        let key = match key_mfp.evaluate(
+                            &mut datums_local,
+                            &temp_storage,
+                            &mut row_packer,
+                        ) {
+                            Err(e) => return Some(Err(DataflowError::from(e))),
+                            Ok(Some(key)) => key,
+                            _ => panic!("Row expected as no predicate was used"),
+                        };
+                        // Evaluate the value expressions.
+                        datums_local.truncate(demand.len());
```

Comment:
i think a comment about why this is needed would be useful

Reply:
Can do.